Reset index after move construction

yuanxingyang opened the pull request:
## Notes for Reviewer
<!-- Items in addition to the checklist below that the reviewer should look for -->

## Pre-Review Checklist for the PR Author

1. [ ] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [ ] Tests follow the [best practice for testing][testing]
1. [ ] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [ ] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [ ] Commits messages are according to this [guideline][commit-guidelines]
    - [ ] Commit messages have the issue ID (`iox-#123 commit text`)
    - [ ] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [ ] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [ ] Relevant issues are linked
1. [ ] Add sensible notes for the reviewer
1. [ ] All checks have passed (except `task-list-completed`)
1. [ ] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/main/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/main/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/main/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [ ] Consider a second reviewer for complex new features or larger refactorings
- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] Unit tests have been written for new behavior
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

<!-- Use either 'Closes #123' or 'Relates to #123' to reference the corresponding issue. -->

- Closes **TBD**

elfenpiff commented:
@yuanxingyang Thank you for your contribution!

Please follow the contribution rules: https://github.com/eclipse-iceoryx/iceoryx/blob/main/CONTRIBUTING.md

1. Sign the eclipse contributor agreement with the same account you committed
2. Create an issue first
3. Adjust every commit message so that it has the prefix `iox-#???` where ??? is the issue number

When this is done I am happy to merge your contribution

elBoberido commented:
@yuanxingyang the process is a bit cumbersome for an initial contributor. The reason for all the 'paperwork' is because we are an Eclipse project (therefore the ECA) and we aim for safety certification (therefore the issue for traceability).

To ease your first contribution, I created #2414 for you. Please use this issue number for your commit. You can change the commit message with `git commit --amend`. For more regular contributions, the git hooks can be used:
https://github.com/eclipse-iceoryx/iceoryx/blob/main/tools/git-hooks/Readme.md

If possible, it would be great to create a tests for the bugfix.

elBoberido commented on iceoryx_hoofs/vocabulary/include/iox/detail/variant.inl:
```diff
@@ -97,6 +97,7 @@ inline constexpr variant<Types...>::variant(variant&& rhs) noexcept
     if (m_type_index != INVALID_VARIANT_INDEX)
     {
         internal::call_at_index<0, Types...>::moveConstructor(m_type_index, &rhs.m_storage, &m_storage);
+        rhs.m_type_index = INVALID_VARIANT_INDEX;
```
Can you check if this also required for the move assignment operator?

yuanxingyang replied:
Hello @elBoberido  
   I have submitted an additional patch. Please review it

elBoberido replied:
@yuanxingyang looks good. Let's see what the CI tells us. Maybe a suppression for a warning accessing a moved object needs to be added.

Did you already signed the ECA? It can take some time until everything is synced.

Can you also please add the issue number to the first commit?

elBoberido replied:
Oh, it seems you need to run clang-format on the tests

elBoberido replied:
... and forget my question regarding the ECA :sweat_smile:

yuanxingyang replied:
> Oh, it seems you need to run clang-format on the tests

I have formatted the latest commit using it

yuanxingyang replied:
> @yuanxingyang looks good. Let's see what the CI tells us. Maybe a suppression for a warning accessing a moved object needs to be added.
> 
> Did you already signed the ECA? It can take some time until everything is synced.
> 
> Can you also please add the issue number to the first commit?

I have forcefully combined two commits into one and added the commit message according to the guidelines.

elBoberido replied:
When the clang-tidy issue is fixed, can you also please add a commit with an entry to the changelog in `iceoryx/doc/website/release-notes/iceoryx-unreleased.md`

yuanxingyang replied:
> When the clang-tidy issue is fixed, can you also please add a commit with an entry to the changelog in `iceoryx/doc/website/release-notes/iceoryx-unreleased.md`

I have add the entry, please check the latest commit

elBoberido commented on iceoryx_hoofs/test/moduletests/test_vocabulary_variant.cpp:
```diff
@@ -352,6 +352,42 @@ TEST_F(variant_Test, MoveCTorWithoutValueResultsInInvalidVariant)
+    iox::variant<int, float, ComplexClass> schlomo;
+    schlomo = ComplexClass(2, 3.14F);
+    iox::variant<int, float, ComplexClass> ignatz(std::move(schlomo));
+    ASSERT_THAT(schlomo.index(), Eq(iox::INVALID_VARIANT_INDEX));
```
```suggestion
    // NOLINTJUSTIFICATION check if move is invalidating the object
    // NOLINTNEXTLINE(bugprone-use-after-move,hicpp-invalid-access-moved,clang-analyzer-cplusplus.Move)
    ASSERT_THAT(schlomo.index(), Eq(iox::INVALID_VARIANT_INDEX));
```
Please add that clang-tidy suppression to all new tests that access the moved from object after the move.

You can also run clang-tidy locally by
```sh
tools/scripts/clang_tidy_check.sh scan_list .clang-tidy-diff-scans.txt iceoryx_hoofs/test/moduletests/test_vocabulary_variant.cpp
```

yuanxingyang replied:
> Please add that clang-tidy suppression to all new tests that access the moved from object after the move.
> 
> You can also run clang-tidy locally by
> 
> ```shell
> tools/scripts/clang_tidy_check.sh scan_list .clang-tidy-diff-scans.txt iceoryx_hoofs/test/moduletests/test_vocabulary_variant.cpp
> ```

The latest commit already includes the changes mentioned above, and running the tool locally did not raise any exceptions regarding my changes

elBoberido commented:
@yuanxingyang it seems some tests are now failing. I guess it's because they were based on assumption which don't hold anymore

yuanxingyang commented:
> @yuanxingyang it seems some tests are now failing. I guess it's because they were based on assumption which don't hold anymore.
> 
> Oh, and please also consider adding your copyright to the files you think have non trivial changes.

In the latest commit, I fixed some issues with the test cases. Since the implementation of EXPECT relies on variant, I also modified some test cases related to EXPECT. Please help review the latest test changes to see if they are reasonable

I am not considering adding my own copyright notice; the modified code is fully shared with everyone

elBoberido commented on iceoryx_hoofs/test/moduletests/test_vocabulary_expected.cpp:
```diff
@@ -340,8 +339,7 @@ TEST_F(expected_test, CreateWithErrorAndMoveCtorLeadsToMovedSource)
     // NOLINTBEGIN(bugprone-use-after-move,hicpp-invalid-access-moved,clang-analyzer-cplusplus.Move)
-    ASSERT_TRUE(sutSource.has_error());
-    EXPECT_TRUE(sutSource.error().m_moved);
+    ASSERT_FALSE(sutSource.has_value());
```
This is a slight semantic change. Can you also add
```cpp
ASSERT_FALSE(sutSource.has_error());
```
In theory, the expected should have neither a value nor an error after move.

yuanxingyang replied:
Fixed in last commit

elBoberido commented on iceoryx_hoofs/test/moduletests/test_vocabulary_expected.cpp:
```diff
@@ -378,8 +375,7 @@ TEST_F(expected_test, CreateWithErrorAndMoveAssignmentLeadsToMovedSource)
     // NOLINTBEGIN(bugprone-use-after-move,hicpp-invalid-access-moved,clang-analyzer-cplusplus.Move)
-    ASSERT_TRUE(sutSource.has_error());
-    EXPECT_TRUE(sutSource.error().m_moved);
+    ASSERT_FALSE(sutSource.has_value());
```
Same here.

yuanxingyang replied:
Fixed in last commit

elBoberido commented on iceoryx_hoofs/test/moduletests/test_vocabulary_variant.cpp:
```diff
@@ -429,12 +471,12 @@ TEST_F(variant_Test, CreatingSecondObjectViaMoveCTorResultsInTwoDTorCalls)
             // NOLINTNEXTLINE(bugprone-use-after-move,hicpp-invalid-access-moved,clang-analyzer-cplusplus.Move)
-            EXPECT_THAT(ignatz.index(), Eq(1U));
+            EXPECT_THAT(ignatz.index(), Eq(iox::INVALID_VARIANT_INDEX));
         }
         EXPECT_THAT(DTorTest::dtorWasCalled, Eq(true));
         DTorTest::dtorWasCalled = false;
     }
-    EXPECT_THAT(DTorTest::dtorWasCalled, Eq(true));
+    EXPECT_THAT(DTorTest::dtorWasCalled, Eq(false));
```
Okay, this leads to a semantic change in the `variant`. The moved from `variant` does not call the d'tor of the underlying value after it is moved. This should be fine though, although maybe unexpected.

@elfenpiff since you wrote the `variant`, your thoughts?

elfenpiff replied:
@elBoberido actually, this is not fine. C++ requires that the dtor is called after an object is moved otherwise we might have a leak here.

Objects must be always safely destructable after move. For instance when one has a threadsafe class, the contents of that class might be moved but since a mutex is unmovable, the mutex is still valid and is only cleaned up after the destructor is called.
This makes sense, since it is allowed to assign a moved class a new value. Simple example could be a threadsafe string. This thing could be moved, and it is legal to assign a completely new string to the moved version.

elBoberido replied:
@yuanxingyang it seems you need to find another solution to the problem you want to fix.

Since the fix was so simple, we did not ask about the problem, which we should have done in the first place. Please excuse this.

Can you describe your issue? With more information we might find a solution that works for you and does not break the contract.

yuanxingyang replied:
**Thank you for your response. I agree with your explanation. Returning to my original question, I made the following modifications:**
`diff --git a/iceoryx_posh/source/runtime/ipc_runtime_interface.cpp b/iceoryx_posh/source/runtime/ipc_runtime_interface.cpp
index ac4d659d7..3e7a3e9a3 100644
--- a/iceoryx_posh/source/runtime/ipc_runtime_interface.cpp
+++ b/iceoryx_posh/source/runtime/ipc_runtime_interface.cpp
@@ -41,6 +41,8 @@ expected<IpcRuntimeInterface, IpcRuntimeInterfaceError> IpcRuntimeInterface::cre
 
     MgmtShmCharacteristics mgmtShmCharacteristics;
 
+    iox::variant<IpcInterfaceUser> tmp;
+    tmp = IpcInterfaceUser(roudi::IPC_CHANNEL_ROUDI_NAME, domainId, ResourceType::ICEORYX_DEFINED);
     auto roudiIpcInterface = IpcInterfaceUser(roudi::IPC_CHANNEL_ROUDI_NAME, domainId, ResourceType::ICEORYX_DEFINED);
`
**When running ./iox-roudi and ./iox-cpp-publisher, the ./iox-cpp-publisher will crash. The backtrace is as follows:"**

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./iox-cpp-publisher...
(gdb) run
Starting program: /home/ynx3sgh/work/projects/iceoryx/build/install/prefix/bin/iox-cpp-publisher 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
2025-01-21 16:41:44.770 [Warn ]: IPC channel still there, doing an unlink of 'iox1_0_u_iox-cpp-publisher'

Program received signal SIGSEGV, Segmentation fault.
0x00005555555ad587 in iox::internal::call_at_index<0ul, iox::runtime::IpcInterfaceUser>::destructor (index=0, ptr=0x7fffffec8810) at /home/ynx3sgh/work/projects/iceoryx/iceoryx_hoofs/vocabulary/include/iox/detail/variant_internal.hpp:177
177	            reinterpret_cast<T*>(ptr)->~T();
(gdb) bt
#0  0x00005555555ad587 in iox::internal::call_at_index<0ul, iox::runtime::IpcInterfaceUser>::destructor (index=0, ptr=0x7fffffec8810)
    at /home/ynx3sgh/work/projects/iceoryx/iceoryx_hoofs/vocabulary/include/iox/detail/variant_internal.hpp:177
#1  0x00005555555acad5 in iox::variant<iox::runtime::IpcInterfaceUser>::call_element_destructor (this=0x7fffffec8810)
    at /home/ynx3sgh/work/projects/iceoryx/iceoryx_hoofs/vocabulary/include/iox/detail/variant.inl:141
#2  0x00005555555ac0e0 in iox::variant<iox::runtime::IpcInterfaceUser>::~variant (this=0x7fffffec8810, __in_chrg=<optimised out>)
    at /home/ynx3sgh/work/projects/iceoryx/iceoryx_hoofs/vocabulary/include/iox/detail/variant.inl:133
#3  0x00005555555aaf40 in iox::runtime::IpcRuntimeInterface::create (runtimeName=..., domainId=..., roudiWaitingTimeout=...)
    at /home/ynx3sgh/work/projects/iceoryx/iceoryx_posh/source/runtime/ipc_runtime_interface.cpp:157
#4  0x000055555557f179 in operator() (__closure=0x7ffffff968f0) at /home/ynx3sgh/work/projects/iceoryx/iceoryx_posh/source/runtime/posh_runtime_impl.cpp:64
#5  0x000055555557f7d2 in iox::runtime::PoshRuntimeImpl::PoshRuntimeImpl (this=0x5555556768c0 <iox::runtime::PoshRuntime::defaultRuntimeFactory(iox::optional<iox::string<87ul> const*>)::buf>, name=..., 
    domainId=..., location=iox::runtime::RuntimeLocation::SEPARATE_PROCESS_FROM_ROUDI) at /home/ynx3sgh/work/projects/iceoryx/iceoryx_posh/source/runtime/posh_runtime_impl.cpp:116
#6  0x000055555557cb6c in operator()<iox::optional<iox::string<87> const*> > (__closure=0x7fffffffd1cf, name=...) at /home/ynx3sgh/work/projects/iceoryx/iceoryx_posh/source/runtime/posh_runtime.cpp:87
#7  0x000055555557cc30 in iox::runtime::PoshRuntime::defaultRuntimeFactory (name=...) at /home/ynx3sgh/work/projects/iceoryx/iceoryx_posh/source/runtime/posh_runtime.cpp:90
#8  0x000055555557cda6 in iox::runtime::PoshRuntime::getInstance (name=...) at /home/ynx3sgh/work/projects/iceoryx/iceoryx_posh/source/runtime/posh_runtime.cpp:107
#9  0x000055555557cd35 in iox::runtime::PoshRuntime::initRuntime (name=...) at /home/ynx3sgh/work/projects/iceoryx/iceoryx_posh/source/runtime/posh_runtime.cpp:102
#10 0x00005555555613db in main () at /home/ynx3sgh/work/projects/iceoryx/iceoryx_examples/icedelivery/iox_publisher.cpp:37
(gdb) quit

yuanxingyang replied:
I probably used the wrong method to assign a value to the variant. When I replaced = with emplace to assign a value, the crash no longer occurred.

elBoberido replied:
I'll have a closer look at it later but if you have time, can you check whether `IpcInterfaceUser` move assignment and dtor is implemented correctly?
This is one of the oldests classes with it's roots in an R&D department and it might be that this class does not implement move correctly or should not be moveable at all.

elBoberido replied:
@yuanxingyang I have the feeling that the rule of 5 was not followed and `IpcInterfaceUser` should not be cooyable at all. This is quite an old class, with it's origins at an R&D department. I'll have a look at it later on.

elBoberido replied:
@yuanxingyang I had a closer look and as it turned out, the converting assignment operator is broken and I'm not sure quite sure if it can easily be fixed.

The problem is that currently the assumption is that if the type index matches the type of the value which is assigned, the `variant` already contains a valid value and can be assigned. In your case, the `variant` does not yet contain a valid value and the assignment operator is called on an uninitialized value. This leads to the issue you described.

I think one needs to have a close look on the semantics of the `std::variant` and then closely mimic that behavior. For example, the `std::variant` does not allow to use the conversion operator if it contains the same type twice.

The problem is that something like this also does not work universally, since `emplace` requires the availability of the move ctor.
```cpp
    if (!has_bad_variant_element_access<T>())
    {
        auto storage = reinterpret_cast<T*>(&m_storage);
        *storage = std::forward<T>(rhs);
    }
    else
    {
        emplace<T>(std::forward<T>(rhs));
    }
```

If you have an idea, feel free to share it.

yuanxingyang replied:
hello @elBoberido 
The latest commit align the behavior with std::variant. The reason is that reinterpret_cast is intended for use on already existing objects, so for creating a new object for the first time, placement new is used

elBoberido commented on iceoryx_hoofs/test/moduletests/test_vocabulary_variant.cpp:
```diff
@@ -450,13 +492,15 @@ TEST_F(variant_Test, CreatingSecondObjectViaMoveAssignmentResultsInTwoDTorCalls)
             // NOLINTNEXTLINE(bugprone-use-after-move,hicpp-invalid-access-moved,clang-analyzer-cplusplus.Move)
-            EXPECT_THAT(ignatz.index(), Eq(1U));
-            EXPECT_THAT(DTorTest::dtorWasCalled, Eq(false));
+            // Reset index after move construction
+            EXPECT_THAT(ignatz.index(), Eq(iox::INVALID_VARIANT_INDEX));
```
Having the comment here will prevent the suppression from working.

yuanxingyang replied:
Fixed in last commit

elBoberido commented on iceoryx_hoofs/vocabulary/include/iox/detail/variant.inl:
```diff
@@ -149,25 +149,22 @@ template <typename T>
-        error_message(__PRETTY_FUNCTION__,
-                      "wrong variant type assignment, another type is already "
-                      "set in variant");
+        if (m_type_index != INVALID_VARIANT_INDEX)
+        {
```
Isn't this if-statement superfluous?

elBoberido commented on iceoryx_hoofs/test/moduletests/test_vocabulary_variant.cpp:
```diff
@@ -473,7 +473,7 @@ TEST_F(variant_Test, DirectValueAssignmentWhenAlreadyAssignedWithDifferentType)
     schlomo = 123;
     schlomo = 123.01F;
-    EXPECT_THAT(schlomo.index(), Eq(0U));
+    EXPECT_THAT(schlomo.index(), Eq(1U));
```
Can you please adjust the name of the test? It currently does not follow our naming convention of"What-Condition-Result". This should become "DirectValueAssignmentWhenAlreadyAssignedWithDifferentTypeResultsInCorrectIndex".

Please also change the test-id, since the semantics of the tests changed. The id can be obtained with `uuidgen`.

yuanxingyang replied:
hello @elBoberido 

 The issues you pointed out above have been fixed.

elBoberido reviewed (commented):
Sorry for letting you wait so long. This totally slipped my attention.

elBoberido reviewed (commented):
@yuanxingyang thanks for you patience. Looks good now. I just want to double check with @elfenpiff if he also agrees with the solution.

elBoberido commented:
@yuanxingyang can you please fix the clang-tidy warning

yuanxingyang commented:
> @yuanxingyang can you please fix the clang-tidy warning
OK， I'll try to fix it

yuanxingyang commented on iceoryx_hoofs/vocabulary/include/iox/detail/variant.inl:
```diff
@@ -149,25 +149,21 @@ template <typename T>
+        call_element_destructor();
+        // NOLINTJUSTIFICATION clang-tidy suggests that std::move(rhs) might erroneously move an lvalue, but here rhs is already T&&, so it's safe.
+        // NOLINTNEXTLINE(bugprone-move-forwarding-reference)
+        new (&m_storage) T(std::move(rhs));
```
> @yuanxingyang Does the suggestion from clang-tidy to use std::forward not work?

It does not work; the following error occurs during compilation:
`In file included from /home/ynx3sgh/work/projects/iceoryx/iceoryx_hoofs/vocabulary/include/iox/variant.hpp:331,
                 from /home/ynx3sgh/work/projects/iceoryx/iceoryx_hoofs/test/moduletests/test_vocabulary_variant.cpp:18:
/home/ynx3sgh/work/projects/iceoryx/iceoryx_hoofs/vocabulary/include/iox/detail/variant.inl: In instantiation of ‘typename std::enable_if<(! std::is_same<T, iox::variant<Types>&>::value), iox::variant<Types> >::type& iox::variant<Types>::operator=(T&&) [with T = int; Types = {int, float}; typename std::enable_if<(! std::is_same<T, iox::variant<Types>&>::value), iox::variant<Types> >::type = iox::variant<int, float>]’:
/home/ynx3sgh/work/projects/iceoryx/iceoryx_hoofs/test/moduletests/test_vocabulary_variant.cpp:466:15:   required from here
/home/ynx3sgh/work/projects/iceoryx/iceoryx_hoofs/vocabulary/include/iox/detail/variant.inl:157:40: error: no matching function for call to ‘forward(int&)’
  157 |         new (&m_storage) T(std::forward(rhs));
`

elBoberido replied:
@yuanxingyang move cannot be used in this context. This operator is used for copy assignment and it would be a bug if a complex type would be assigned to a default constructed variant.

Unfortunately, our tests currently do not cover this scenario yet.
